File: coptoc/ios/Sources/Subjects.swift
import SwiftUI

/// §5.6b the subject of concern on the phone. The wall opens files, rates them and works the inbox; the phone reads
/// the files and the inbox, shows on a principal who is directed at them, and files a contact from the field — the
/// in-person door of the mailroom (someone approached the principal at the venue; a letter was handed to the desk).
/// Nothing here rates a person: that is the S2 analyst's, at the wall. Everything is empty unless the signed-in role
/// is cleared for the files; the tally in the summary is what an uncleared reader sees.

private let RATING_COLOR: [String: Color] = ["green": Theme.green, "amber": Theme.amber, "red": Theme.red]
private func ratingColor(_ r: String) -> Color { RATING_COLOR[r] ?? Theme.line }
private let DIRECT_LABEL: [String: String] = ["directed": "DIRECTED", "conditional": "CONDITIONAL", "veiled": "VEILED", "none": "NO THREAT"]
private func directColor(_ d: String) -> Color { d == "directed" ? Theme.red : d == "conditional" ? Theme.orange : d == "veiled" ? Theme.amber : Theme.dim }
private func ago(_ d: Double?) -> String { guard let d else { return "never assessed" }; return d < 1 ? "today" : "\(Int(d.rounded()))d ago" }

/// The row of ratings, one block per indicator in the configured order.
struct RatingBlocks: View {
    var strip: [String]
    var body: some View {
        HStack(spacing: 2) { ForEach(Array(strip.enumerated()), id: \.offset) { _, r in RoundedRectangle(cornerRadius: 2).fill(ratingColor(r)).frame(width: 8, height: 8) } }
    }
}

/// What a principal or a site carries about a file: the name, the strip, the worst of it as a word.
struct SubjectStripRow: View {
    @Environment(COPStore.self) private var store
    var s: SubjectCompact
    var body: some View {
        Button { store.selection = .subject(s.id) } label: {
            HStack(spacing: 6) {
                Text(s.name).font(.system(size: 12, weight: .semibold)).lineLimit(1)
                RatingBlocks(strip: s.strip)
                if s.unassessed { Chip(text: "UNASSESSED", color: Theme.amber) } else { Chip(text: s.worst.uppercased(), color: ratingColor(s.worst), filled: s.worst == "red") }
                Spacer()
                if s.contactCount > 0 { Text("\(s.contactCount) contact\(s.contactCount == 1 ? "" : "s")").font(.system(size: 10, design: .monospaced)).foregroundStyle(Theme.dim) }
                if s.stale { Chip(text: "STALE", color: Theme.amber) }
            }
        }.buttonStyle(.plain)
    }
}

/// The S2 tab section: every live file worst first, then what nobody has attributed yet.
struct SubjectsSection: View {
    @Environment(COPStore.self) private var store
    var cleared: Bool { ["battle_captain", "ep", "analyst"].contains(store.snapshot?.me?.role ?? store.client.role) }
    var body: some View {
        let live = store.liveSubjects
        let rest = (store.snapshot?.subjects ?? []).filter { !$0.live }
        let inbox = store.subjectInbox
        let tally = store.snapshot?.summary
        Section(header: HStack { SectionLabel(text: "WHO IS DIRECTED AT US · \(tally?.subjectsOpen ?? live.count) OPEN · \(tally?.subjectsRed ?? live.filter { $0.worst == "red" }.count) RED"); Spacer(); if cleared { ContactButton(subjectId: nil, compact: true) } }) {
            if !cleared {
                Text("Subject files name private individuals. Battle Captain, Executive Protection or the S2 analyst only.").font(.system(size: 11)).foregroundStyle(Theme.dim)
            } else {
                ForEach(live + rest) { s in
                    Button { store.selection = .subject(s.id) } label: {
                        VStack(alignment: .leading, spacing: 3) {
                            HStack(spacing: 6) {
                                Text(s.name).font(.system(size: 12, weight: .semibold)).lineLimit(1)
                                if s.unassessed { Chip(text: "UNASSESSED", color: Theme.amber) } else { Chip(text: s.worst.uppercased(), color: ratingColor(s.worst), filled: s.worst == "red") }
                                if s.status != "open" { Chip(text: s.status.uppercased(), color: s.status == "monitoring" ? Theme.blue : s.status == "referred" ? Theme.gold : Theme.dim) }
                                Spacer()
                            }
                            HStack(spacing: 6) {
                                RatingBlocks(strip: (s.assessment?.ratings ?? []).map(\.rating))
                                Text([s.principalName.isEmpty ? (s.worstIndicator ?? "nothing rated") : "directed at \(s.principalName)",
                                      s.contactCount > 0 ? "\(s.contactCount) contact\(s.contactCount == 1 ? "" : "s")" : nil,
                                      s.worstDirectness != "none" ? DIRECT_LABEL[s.worstDirectness]?.lowercased() : nil,
                                      s.assessedAt == nil ? "never assessed" : ago(s.ageDays)].compactMap { $0 }.joined(separator: " · "))
                                    .font(.system(size: 10, design: .monospaced)).foregroundStyle(Theme.dim).lineLimit(1)
                            }
                        }.opacity(s.live ? 1 : 0.55)
                    }.buttonStyle(.plain)
                }
                if live.isEmpty && rest.isEmpty { Text("No file is open on anyone.").font(.system(size: 11)).foregroundStyle(Theme.dim) }
                if !inbox.isEmpty {
                    SectionLabel(text: "INBOX · \(inbox.count) ATTRIBUTED TO NOBODY")
                    ForEach(inbox) { c in
                        VStack(alignment: .leading, spacing: 3) {
                            HStack(spacing: 6) {
                                Chip(text: c.channel.uppercased().replacingOccurrences(of: "_", with: " "), color: Theme.dim)
                                Text(c.fromLabel.isEmpty ? "sender not given" : c.fromLabel).font(.system(size: 11, weight: .semibold)).lineLimit(1)
                                Spacer()
                                Chip(text: DIRECT_LABEL[c.directness] ?? c.directness.uppercased(), color: directColor(c.directness), filled: c.directness == "directed")
                            }
                            Text(c.text).font(.system(size: 11)).foregroundStyle(.secondary).lineLimit(3)
                            Text("\(ISO.rel(c.receivedAt, now: store.now))" + (c.principalName.isEmpty ? " · names nobody we hold" : " · names \(c.principalName)") + " · attributed at the wall")
                                .font(.system(size: 10, design: .monospaced)).foregroundStyle(Theme.dim).lineLimit(1)
                        }
                    }
                }
            }
        }.listRowBackground(Theme.panel)
    }
}

/// The detail card for one file: who he is, the assessment as the analyst wrote it, what has arrived from him.
struct SubjectDetail: View {
    @Environment(COPStore.self) private var store
    var s: Subject
    func kicker(_ t: String) -> some View { Text(t).font(.system(size: 10, design: .monospaced)).tracking(1.4).foregroundStyle(Theme.dim) }
    func kv(_ k: String, _ v: String) -> some View { HStack(alignment: .top, spacing: 10) { Text(k).foregroundStyle(Theme.dim).frame(width: 64, alignment: .leading); Text(v) }.font(.system(size: 12)) }
    var body: some View {
        kicker("SUBJECT OF CONCERN · RESTRICTED · \(s.status.uppercased())" + (s.assessment.map { " · \($0.assessedBy)" } ?? ""))
        HStack(spacing: 8) {
            Text(s.name).font(.system(size: 20, weight: .bold))
            if s.unassessed { Chip(text: "NEVER ASSESSED", color: Theme.amber) } else { Chip(text: s.worst.uppercased(), color: ratingColor(s.worst), filled: s.worst == "red") }
            if s.stale { Chip(text: "STALE", color: Theme.amber) }
        }
        if !s.aliases.isEmpty { Text("also: " + s.aliases.joined(separator: " · ")).font(.system(size: 11)).foregroundStyle(Theme.dim) }
        if !s.summary.isEmpty { Text(s.summary).font(.system(size: 12)).foregroundStyle(.secondary) }
        HStack(spacing: 8) {
            if let pid = s.principalId { Button("⌖ \(s.principalName)") { store.selection = .person(pid) }.font(.system(size: 11, weight: .semibold, design: .monospaced)).buttonStyle(.bordered).tint(Theme.blue) }
            if let lid = s.locationId, store.site(lid) != nil { Button("⌖ SITE") { store.selection = .site(lid) }.font(.system(size: 11, weight: .semibold, design: .monospaced)).buttonStyle(.bordered).tint(Theme.blue) }
        }
        if s.status == "referred", let r = s.referredTo { kv("Referred", r) }
        if s.status == "closed", let r = s.closedReason { kv("Closed", r) }
        if let seen = s.lastSeenPlace { kv("Last seen", seen + (s.lastSeenAt.map { " · \(ISO.rel($0, now: store.now))" } ?? "")) }

        SectionLabel(text: "THE ASSESSMENT" + (s.unassessed ? " · NONE YET" : " · \(ago(s.ageDays)) · nothing summed"))
        if s.unassessed {
            Text("Nobody has rated this person. That is an exception in its own right; it is not a green file. Rating is done at the wall.").font(.system(size: 11)).foregroundStyle(Theme.dim)
        } else if let a = s.assessment {
            ForEach(a.ratings, id: \.indicator) { r in
                HStack(alignment: .top, spacing: 8) {
                    Rectangle().fill(ratingColor(r.rating)).frame(width: 3, height: 30)
                    VStack(alignment: .leading, spacing: 2) {
                        HStack(spacing: 6) { Text(r.label).font(.system(size: 12, weight: .semibold)); Chip(text: r.rating == "unknown" ? "—" : r.rating.uppercased(), color: ratingColor(r.rating)) }
                        Text(r.note.isEmpty ? "no justification recorded" : r.note).font(.system(size: 11)).foregroundStyle(r.note.isEmpty ? Theme.dim : .secondary).lineLimit(3)
                    }
                }
            }
            if !a.summary.isEmpty { Text(a.summary).font(.system(size: 12)).foregroundStyle(.secondary).padding(.top, 2) }
        }

        HStack { SectionLabel(text: "WHAT HAS ARRIVED · \(s.contactCount)"); Spacer(); if s.live { ContactButton(subjectId: s.id, compact: true) } }
        if s.contacts.isEmpty { Text("Nothing has been attributed to this file.").font(.system(size: 11)).foregroundStyle(Theme.dim) }
        ForEach(s.contacts) { c in
            VStack(alignment: .leading, spacing: 3) {
                HStack(spacing: 6) {
                    Chip(text: c.channel.uppercased().replacingOccurrences(of: "_", with: " "), color: Theme.dim)
                    Text(c.fromLabel.isEmpty ? "sender not given" : c.fromLabel).font(.system(size: 11, weight: .semibold)).lineLimit(1)
                    Spacer()
                    Chip(text: DIRECT_LABEL[c.directness] ?? c.directness.uppercased(), color: directColor(c.directness), filled: c.directness == "directed")
                    Chip(text: "\(c.reliability)/\(c.credibility)", color: Theme.dim)
                }
                Text(c.text).font(.system(size: 11)).foregroundStyle(.secondary)
                Text("\(ISO.rel(c.receivedAt, now: store.now))" + (c.principalName.isEmpty ? "" : " · names \(c.principalName)") + (c.note.map { " · \($0)" } ?? ""))
                    .font(.system(size: 10, design: .monospaced)).foregroundStyle(Theme.dim).lineLimit(2)
            }.padding(.vertical, 2)
        }
        Text("Opened by \(s.openedBy) · \(ISO.rel(s.openedAt, now: store.now))" + (s.caseId.map { " · case \($0)" } ?? "") + ". Rated, referred and closed at the wall.").font(.system(size: 10, design: .monospaced)).foregroundStyle(Theme.dim).padding(.top, 4)
    }
}

// ---------------------------------------------------------------- the field door of the mailroom

struct ContactButton: View {
    var subjectId: String?
    var compact: Bool
    @State private var open = false
    var body: some View {
        Button { open = true } label: {
            HStack(spacing: 5) { Text("+").font(.system(size: compact ? 10 : 14, weight: .heavy)); Text("CONTACT").font(.system(size: compact ? 9 : 11, weight: .bold, design: .monospaced)).tracking(1) }
                .foregroundStyle(.white).padding(.horizontal, compact ? 8 : 12).padding(.vertical, compact ? 4 : 9)
                .background(Theme.blue.opacity(compact ? 0.35 : 0.9), in: RoundedRectangle(cornerRadius: 8))
                .overlay(RoundedRectangle(cornerRadius: 8).stroke(Theme.blue, lineWidth: 1))
        }
        .buttonStyle(.plain)
        .sheet(isPresented: $open) { ContactForm(subjectId: subjectId) { open = false } }
    }
}

/// Something arrived, or someone approached. It lands graded F/6 — an unknown sender cannot be judged — and, unless
/// filed from a subject's own card, in the inbox: the phone never guesses whose it is.
struct ContactForm: View {
    @Environment(COPStore.self) private var store
    var subjectId: String?
    var onDone: () -> Void
    @State private var channel = "in_person"
    @State private var from = ""; @State private var text = ""; @State private var directness = "none"; @State private var principal = ""
    @State private var problem: String? = nil
    private let channels = [("in_person", "IN PERSON"), ("phone", "PHONE"), ("letter", "LETTER"), ("email", "EMAIL"), ("dm", "DM"), ("form", "FORM"), ("other", "OTHER")]
    var subject: Subject? { store.subject(subjectId) }
    var body: some View {
        NavigationStack {
            Form {
                Section(subject.map { "Onto the file: \($0.name)" } ?? "Into the inbox — attributed at the wall") {
                    Picker("Channel", selection: $channel) { ForEach(channels, id: \.0) { Text($0.1).tag($0.0) } }
                    TextField(channel == "in_person" ? "Who — as they gave it, or a description" : "From — address, handle or return address as it arrived", text: $from)
                    TextField("What was said or written, in the words it arrived in", text: $text, axis: .vertical).lineLimit(3...8)
                }
                Section("How the threat is worded — the wording, not your view of the risk") {
                    Picker("Threat", selection: $directness) { Text("NO THREAT").tag("none"); Text("VEILED").tag("veiled"); Text("CONDITIONAL").tag("conditional"); Text("DIRECTED").tag("directed") }.pickerStyle(.segmented)
                }
                if subject == nil {
                    Section("Who it names, if one of ours") {
                        Picker("Principal", selection: $principal) { Text("— nobody named —").tag(""); ForEach((store.snapshot?.people ?? []).filter(\.isVip)) { Text($0.name).tag($0.id) } }
                    }
                }
                Section { Text("Filed by \(store.me?.name ?? store.client.actor) · graded F/6 on arrival — an unknown sender cannot be judged · time is now").font(.system(size: 11)).foregroundStyle(Theme.dim) }
                if let problem { Text(problem).font(.system(size: 12)).foregroundStyle(Theme.red) }
            }
            .navigationTitle("CONTACT")
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel") { onDone() } }
                ToolbarItem(placement: .confirmationAction) { Button("File") { file() }.disabled(text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty) }
            }
        }
    }
    private func file() {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !t.isEmpty else { problem = "Say what arrived, in the words it arrived in."; return }
        var body: [String: Any] = ["text": t, "channel": channel, "from_label": from.trimmingCharacters(in: .whitespacesAndNewlines), "directness": directness,
                                   "received_by": store.me?.name ?? store.client.actor]
        if let subjectId { body["subject_id"] = subjectId }
        if !principal.isEmpty { body["principal_id"] = principal }
        store.act("filing the contact") { try await store.client.fileContact(body) }
        onDone()
    }
}

File: coptoc/android/app/src/main/kotlin/com/toc/coptoc/Subjects.kt
package com.toc.coptoc

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** §5.6b the subject of concern on the phone. The wall opens files, rates them and works the inbox; the phone reads
 *  the files and the inbox, shows on a principal who is directed at them, and files a contact from the field — the
 *  in-person door of the mailroom. It never rates a person: that is the S2 analyst's, at the wall. Everything here is
 *  empty unless the signed-in role is cleared for the files; the tally in the summary is what the floor sees. */

val SUBJECT_READERS = listOf("battle_captain", "ep", "analyst")
fun ratingColor(r: String): Color = when (r) { "green" -> Palette.green; "amber" -> Palette.amber; "red" -> Palette.red; else -> Palette.line }
fun directLabel(d: String) = when (d) { "directed" -> "DIRECTED"; "conditional" -> "CONDITIONAL"; "veiled" -> "VEILED"; else -> "NO THREAT" }
fun directColor(d: String): Color = when (d) { "directed" -> Palette.red; "conditional" -> Palette.orange; "veiled" -> Palette.amber; else -> Palette.dim }
private fun ago(d: Double?): String = if (d == null) "never assessed" else if (d < 1) "today" else "${Math.round(d)}d ago"
private fun channelLabel(c: String) = c.uppercase().replace('_', ' ')

/** The row of ratings, one block per indicator in the configured order. */
@Composable fun RatingBlocks(strip: List<String>) = Row(horizontalArrangement = Arrangement.spacedBy(2.dp)) { strip.forEach { Box(Modifier.size(8.dp).background(ratingColor(it), RoundedCornerShape(2.dp))) } }

@Composable private fun WorstChip(unassessed: Boolean, worst: String) =
    if (unassessed) Chip("UNASSESSED", Palette.amber) else Chip(worst.uppercase(), ratingColor(worst), filled = worst == "red")

/** What a principal or a site carries about a file: the name, the strip, the worst of it as a word. */
@Composable fun SubjectStripRow(s: SubjectCompact, onClick: () -> Unit) =
    Row(Modifier.fillMaxWidth().clickable { onClick() }.padding(vertical = 2.dp), horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
        Text(s.name, color = Palette.text, fontSize = 11.5.sp, fontWeight = FontWeight.SemiBold, maxLines = 1, overflow = TextOverflow.Ellipsis)
        RatingBlocks(s.strip)
        WorstChip(s.unassessed, s.worst)
        Spacer(Modifier.weight(1f))
        if (s.contactCount > 0) Text("${s.contactCount} contact" + (if (s.contactCount == 1) "" else "s"), color = Palette.dim, fontSize = 9.sp, fontFamily = FontFamily.Monospace)
        if (s.stale) Chip("STALE", Palette.amber)
    }

/** The S2 tab section: every live file worst first, then what nobody has attributed yet. */
fun LazyListScope.subjectsSection(st: WallState, store: Store, onFile: () -> Unit) {
    val snap = st.snap ?: return
    val cleared = (snap.me?.role ?: st.role) in SUBJECT_READERS
    val live = snap.subjects.filter { it.live }; val rest = snap.subjects.filter { !it.live }
    val t = snap.summary
    item { Label("WHO IS DIRECTED AT US", "${t.subjectsOpen} open · ${t.subjectsRed} red", action = { if (cleared) Mini("+ CONTACT", Palette.blue2, st.busy == null) { onFile() } }) }
    if (!cleared) { item { Text("Subject files name private individuals. Battle Captain, Executive Protection or the S2 analyst only.", Modifier.padding(horizontal = 14.dp), color = Palette.dim, fontSize = 10.sp) }; return }
    items(live + rest, key = { "subj_" + it.id }) { s ->
        RowItem(selected = (st.selection as? Selection.SubjectSel)?.id == s.id, onClick = { store.select(Selection.SubjectSel(s.id)) }) {
            Column(Modifier.weight(1f).alpha(if (s.live) 1f else .55f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                    Text(s.name, color = Palette.text, fontSize = 11.5.sp, fontWeight = FontWeight.SemiBold, maxLines = 1, overflow = TextOverflow.Ellipsis)
                    WorstChip(s.unassessed, s.worst)
                    if (s.status != "open") Chip(s.status.uppercase(), when (s.status) { "monitoring" -> Palette.blue2; "referred" -> Palette.amber; else -> Palette.dim }) }
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                    RatingBlocks(s.assessment?.ratings?.map { it.rating } ?: emptyList())
                    Text(listOfNotNull(if (s.principalName.isNotEmpty()) "directed at ${s.principalName}" else (s.worstIndicator ?: "nothing rated"),
                                       if (s.contactCount > 0) "${s.contactCount} contact" + (if (s.contactCount == 1) "" else "s") else null,
                                       if (s.worstDirectness != "none") directLabel(s.worstDirectness).lowercase() else null,
                                       if (s.assessedAt == null) "never assessed" else ago(s.ageDays)).joinToString(" · "),
                         color = Palette.dim, fontSize = 9.sp, fontFamily = FontFamily.Monospace, maxLines = 1, overflow = TextOverflow.Ellipsis) } } } }
    if (live.isEmpty() && rest.isEmpty()) item { Text("No file is open on anyone.", Modifier.padding(horizontal = 14.dp), color = Palette.dim, fontSize = 10.sp) }
    if (snap.subjectInbox.isNotEmpty()) {
        item { Label("INBOX", "${snap.subjectInbox.size} attributed to nobody") }
        items(snap.subjectInbox, key = { "inbox_" + it.id }) { c ->
            Column(Modifier.padding(horizontal = 14.dp, vertical = 6.dp), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                    Chip(channelLabel(c.channel)); Text(c.fromLabel.ifEmpty { "sender not given" }, color = Palette.text, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.weight(1f))
                    Chip(directLabel(c.directness), directColor(c.directness), filled = c.directness == "directed") }
                Text(c.text, color = Palette.text.copy(alpha = .85f), fontSize = 10.5.sp, maxLines = 3, overflow = TextOverflow.Ellipsis)
                Text(c.receivedAt.take(16).replace('T', ' ') + "Z · " + (if (c.principalName.isEmpty()) "names nobody we hold" else "names ${c.principalName}") + " · attributed at the wall", color = Palette.dim, fontSize = 9.sp, fontFamily = FontFamily.Monospace, maxLines = 1, overflow = TextOverflow.Ellipsis)
                HorizontalDivider(thickness = 0.5.dp, color = Palette.line) } } }
}

/** The detail card for one file: who he is, the assessment as the analyst wrote it, what has arrived from him. */
@Composable fun ColumnScope.SubjectDetailBody(s: Subject, st: WallState, store: Store, onFile: () -> Unit) {
    val snap = st.snap ?: return
    Kicker("SUBJECT OF CONCERN · RESTRICTED · ${s.status.uppercase()}" + (s.assessment?.let { " · ${it.assessedBy}" } ?: ""))
    Row(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) { Title(s.name); WorstChip(s.unassessed, s.worst); if (s.stale) Chip("STALE", Palette.amber) }
    if (s.aliases.isNotEmpty()) Text("also: " + s.aliases.joinToString(" · "), color = Palette.dim, fontSize = 10.5.sp)
    if (s.summary.isNotEmpty()) Text(s.summary, color = Palette.text.copy(alpha = .9f), fontSize = 11.sp, lineHeight = 15.sp)
    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        s.principalId?.let { pid -> Mini("⌖ ${s.principalName}", Palette.blue2) { store.select(Selection.PersonSel(pid)) } }
        s.locationId?.let { lid -> if (snap.locations.any { it.id == lid }) Mini("⌖ SITE", Palette.blue2) { store.select(Selection.SiteSel(lid)) } } }
    if (s.status == "referred") s.referredTo?.let { KV("Referred", it) }
    if (s.status == "closed") s.closedReason?.let { KV("Closed", it) }
    s.lastSeenPlace?.let { KV("Last seen", it + (s.lastSeenAt?.let { t -> " · " + t.take(16).replace('T', ' ') + "Z" } ?: "")) }

    Section("THE ASSESSMENT", if (s.unassessed) "none yet" else ago(s.ageDays) + " · nothing summed")
    if (s.unassessed) Text("Nobody has rated this person. That is an exception in its own right; it is not a green file. Rating is done at the wall.", color = Palette.dim, fontSize = 10.sp)
    else s.assessment?.let { a ->
        a.ratings.forEach { r -> Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.Top) {
            Box(Modifier.width(3.dp).height(30.dp).background(ratingColor(r.rating)))
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) { Text(r.label, color = Palette.text, fontSize = 11.5.sp, fontWeight = FontWeight.SemiBold); Chip(if (r.rating == "unknown") "—" else r.rating.uppercase(), ratingColor(r.rating)) }
                Text(r.note.ifEmpty { "no justification recorded" }, color = if (r.note.isEmpty()) Palette.dim else Palette.text.copy(alpha = .8f), fontSize = 10.5.sp, maxLines = 3, overflow = TextOverflow.Ellipsis) } } }
        if (a.summary.isNotEmpty()) Text(a.summary, color = Palette.text.copy(alpha = .85f), fontSize = 11.sp, lineHeight = 15.sp) }

    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) { Section("WHAT HAS ARRIVED", "${s.contactCount}"); Spacer(Modifier.weight(1f)); if (s.live) Mini("+ CONTACT", Palette.blue2, st.busy == null) { onFile() } }
    if (s.contacts.isEmpty()) Text("Nothing has been attributed to this file.", color = Palette.dim, fontSize = 10.sp)
    s.contacts.forEach { c -> Column(Modifier.padding(vertical = 3.dp), verticalArrangement = Arrangement.spacedBy(3.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
            Chip(channelLabel(c.channel)); Text(c.fromLabel.ifEmpty { "sender not given" }, color = Palette.text, fontSize = 11.sp, fontWeight = FontWeight.SemiBold, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.weight(1f))
            Chip(directLabel(c.directness), directColor(c.directness), filled = c.directness == "directed"); Chip("${c.reliability}/${c.credibility}") }
        Text(c.text, color = Palette.text.copy(alpha = .85f), fontSize = 10.5.sp)
        Text(c.receivedAt.take(16).replace('T', ' ') + "Z" + (if (c.principalName.isEmpty()) "" else " · names ${c.principalName}") + (c.note?.let { " · $it" } ?: ""), color = Palette.dim, fontSize = 9.sp, fontFamily = FontFamily.Monospace, maxLines = 2, overflow = TextOverflow.Ellipsis) } }
    Text("Opened by ${s.openedBy} · ${s.openedAt.take(16).replace('T', ' ')}Z" + (s.caseId?.let { " · case $it" } ?: "") + ". Rated, referred and closed at the wall.", color = Palette.dim, fontSize = 9.sp, fontFamily = FontFamily.Monospace)
}

/** Something arrived, or someone approached. It lands graded F/6 — an unknown sender cannot be judged — and, unless
 *  filed from a subject's own card, in the inbox: the phone never guesses whose it is. */
@Composable fun ContactDialog(st: WallState, store: Store, subjectId: String?, open: Boolean, onDone: () -> Unit) {
    if (!open) return
    val subject = st.snap?.subjects?.firstOrNull { it.id == subjectId }
    var channel by remember { mutableStateOf("in_person") }
    var from by remember { mutableStateOf("") }; var text by remember { mutableStateOf("") }; var directness by remember { mutableStateOf("none") }; var principal by remember { mutableStateOf("") }
    val me = st.snap?.me
    val channels = listOf("in_person" to "IN PERSON", "phone" to "PHONE", "letter" to "LETTER", "email" to "EMAIL", "dm" to "DM", "form" to "FORM", "other" to "OTHER")
    val vips = st.snap?.people?.filter { it.isVip } ?: emptyList()
    AlertDialog(onDismissRequest = onDone, containerColor = Palette.panel, titleContentColor = Palette.text, textContentColor = Palette.text,
        title = { Text(if (subject != null) "CONTACT · ${subject.name}" else "CONTACT · INTO THE INBOX", fontSize = 13.sp, fontFamily = FontFamily.Monospace) },
        text = { Column(verticalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.verticalScroll(rememberScrollState())) {
            Text(if (subject != null) "Onto the file. Rated at the wall." else "Attributed to nobody until an analyst says so, at the wall.", color = Palette.dim, fontSize = 9.5.sp)
            Row(Modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(5.dp)) { channels.forEach { (k, l) -> Chip(l, if (channel == k) Palette.blue2 else Palette.dim, filled = channel == k) { channel = k } } }
            OutlinedTextField(from, { from = it }, label = { Text(if (channel == "in_person") "Who — as they gave it, or a description" else "From — as it arrived") }, singleLine = true)
            OutlinedTextField(text, { text = it }, label = { Text("What was said or written, in the words it arrived in") }, minLines = 3)
            Text("How the threat is worded — the wording, not your view of the risk", color = Palette.dim, fontSize = 9.sp)
            Row(horizontalArrangement = Arrangement.spacedBy(5.dp)) { listOf("none", "veiled", "conditional", "directed").forEach { d -> Chip(directLabel(d), if (directness == d) directColor(d) else Palette.dim, filled = directness == d) { directness = d } } }
            if (subject == null && vips.isNotEmpty()) {
                Text("Who it names, if one of ours", color = Palette.dim, fontSize = 9.sp)
                Row(Modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(5.dp)) { Chip("NOBODY", if (principal.isEmpty()) Palette.blue2 else Palette.dim, filled = principal.isEmpty()) { principal = "" }
                    vips.forEach { p -> Chip(p.name, if (principal == p.id) Palette.blue2 else Palette.dim, filled = principal == p.id) { principal = p.id } } } }
            Text("Filed by ${me?.name ?: store.api.actor} · graded F/6 on arrival — an unknown sender cannot be judged · time is now", color = Palette.dim, fontSize = 9.sp) } },
        confirmButton = { TextButton({
            if (text.isNotBlank()) {
                val body = buildJsonObject { put("text", text.trim()); put("channel", channel); put("from_label", from.trim()); put("directness", directness); put("received_by", me?.name ?: store.api.actor)
                    subjectId?.let { put("subject_id", it) }; if (principal.isNotEmpty()) put("principal_id", principal) }
                store.act("filing the contact") { fileContact(body) }; onDone() } }) { Text("FILE", color = Palette.blue2) } },
        dismissButton = { TextButton(onDone) { Text("CANCEL", color = Palette.dim) } })
}

